src/server/gc_worker/applied_lock_collector.rs

```diff
@@ -0,0 +1,746 @@
+enum LockObserverMsg {
+    // A lock is applied.
+    Locks(Vec<(Key, Lock)>),
+    // A error occurs in the observer,
+    Err(Error),
+}
+
+impl Display for LockObserverMsg {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Debug::fmt(&self, f)
+    }
+}
+
+pub type Callback<T> = Box<dyn FnOnce(Result<T>) + Send>;
+
+enum LockCollectorTask {
+    // Messages from observer
+    ObserverMsg(LockObserverMsg),
+
+    // Messages from client
+    StartCollecting {
+        max_ts: TimeStamp,
+        callback: Callback<()>,
+    },
+    GetCollectedLocks {
+        max_ts: TimeStamp,
+        callback: Callback<(Vec<LockInfo>, bool)>,
+    },
+    StopCollecting {
+        max_ts: TimeStamp,
+        callback: Callback<()>,
+    },
+}
+
+impl Debug for LockCollectorTask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            LockCollectorTask::ObserverMsg(msg) => {
+                f.debug_struct("ObserverMsg").field("msg", msg).finish()
+            }
+            LockCollectorTask::StartCollecting { max_ts, .. } => f
+                .debug_struct("StartCollecting")
+                .field("max_ts", max_ts)
+                .finish(),
+            LockCollectorTask::GetCollectedLocks { max_ts, .. } => f
+                .debug_struct("GetCollectedLocks")
+                .field("max_ts", max_ts)
+                .finish(),
+            LockCollectorTask::StopCollecting { max_ts, .. } => f
+                .debug_struct("StopCollecting")
+                .field("max_ts", max_ts)
+                .finish(),
+        }
+    }
+}
+
+impl Display for LockCollectorTask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Debug::fmt(&self, f)
+    }
+}
+
+/// `LockObserver` observes apply events and apply snapshot events. If it happens in CF_LOCK, it
+/// checks the `start_ts`s of the locks being written. If a lock's `start_ts` <= specified `max_ts`
+/// in the `state`, it will send the lock to through the `sender`, so the receiver can collect it.
+#[derive(Clone)]
+struct LockObserver {
+    state: Arc<LockObserverState>,
+    sender: Scheduler<LockCollectorTask>,
+}
+
+impl LockObserver {
+    pub fn new(state: Arc<LockObserverState>, sender: Scheduler<LockCollectorTask>) -> Self {
+        Self { state, sender }
+    }
+
+    pub fn register(self, coprocessor_host: &mut CoprocessorHost) {
+        coprocessor_host
+            .registry
+            .register_apply_snapshot_observer(1, Box::new(self.clone()));
+        coprocessor_host
+            .registry
+            .register_query_observer(1, Box::new(self));
+    }
+
+    fn send(&self, msg: LockObserverMsg) {
+        match self.sender.schedule(LockCollectorTask::ObserverMsg(msg)) {
+            Ok(()) => (),
+            Err(ScheduleError::Stopped(m)) => {
+                error!("failed to send lock observer msg"; "msg" => ?m);
+            }
+            Err(ScheduleError::Full(m)) => {
```

Comment:
I think it never happens because it's an unbounded channel. If it happens, we should mark `is_clean` as false.